hibernate-core/src/main/java/org/hibernate/engine/spi/LoadQueryInfluencers.java

```diff
@@ -183,18 +187,56 @@ public void disableFetchProfile(String name) throws UnknownProfileException {
 	}
 
 	public EntityGraph getFetchGraph() {
-		return fetchGraph;
+		return entityGraphSemantic == EntityGraphSemantic.FETCH
+				? entityGraph
+				: null;
 	}
 
 	public void setFetchGraph(final EntityGraph fetchGraph) {
-		this.fetchGraph = fetchGraph;
+		if ( fetchGraph == null  ) {
+			throw new IllegalArgumentException( "Passed JPA fetch EntityGraph cannot be null" );
+		}
+
+		this.entityGraphSemantic = EntityGraphSemantic.FETCH;
+		this.entityGraph = fetchGraph;
 	}
 
 	public EntityGraph getLoadGraph() {
-		return loadGraph;
+		return entityGraphSemantic == EntityGraphSemantic.LOAD
+				? entityGraph
+				: null;
 	}
 
 	public void setLoadGraph(final EntityGraph loadGraph) {
-		this.loadGraph = loadGraph;
+		if ( loadGraph == null  ) {
+			throw new IllegalArgumentException( "Passed JPA load EntityGraph cannot be null" );
+		}
+
+		this.entityGraphSemantic = EntityGraphSemantic.LOAD;
+		this.entityGraph = loadGraph;
+	}
+
+	public void applyEntityGraphHint(Map<String, Object> properties) {
+		if ( properties == null || properties.isEmpty() ) {
+			return;
+		}
+
+		final EntityGraph fetchHint = (EntityGraph) properties.get( EntityGraphSemantic.FETCH.getJpaHintName() );
+		if ( fetchHint != null ) {
+			this.entityGraph = fetchHint;
+			this.entityGraphSemantic = EntityGraphSemantic.FETCH;
+		}
+		else {
+			final EntityGraph loadHint = (EntityGraph) properties.get( EntityGraphSemantic.LOAD.getJpaHintName() );
+			if ( loadHint != null ) {
+				this.entityGraph = loadHint;
+				this.entityGraphSemantic = EntityGraphSemantic.FETCH;
```

Comment:
Shouldn't it be `LOAD`?

Reply:
Yep, like I said earlier - still lots of copy/paste things.  In general this is ongoing work, just wanted to get feedback on the overall parts